File: android/app/src/main/java/com/skiftappen/widget/SupabaseService.kt
package com.skiftappen.widget

import android.content.Context
import android.content.SharedPreferences
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * Service for communicating with Supabase API from Android widget
 */
class SupabaseService private constructor(private val context: Context) {
    
    companion object {
        private const val SUPABASE_URL = "https://fsefeherdbtsddqimjco.supabase.co"
        private const val SUPABASE_ANON_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImZzZWZlaGVyZGJ0c2RkcWltamNvIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NTI3ODUwNDcsImV4cCI6MjA2ODM2MTA0N30.YEltOJVQU6Ox5YrkZJGzbMiojyQClkFwG-mBPilIAfk"
        private const val PREFS_NAME = "skiftappen_widget_prefs"
        private const val KEY_ACCESS_TOKEN = "supabase_access_token"
        private const val KEY_USER_ID = "user_id"
        
        @Volatile
        private var INSTANCE: SupabaseService? = null
        
        fun getInstance(context: Context): SupabaseService {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: SupabaseService(context.applicationContext).also { INSTANCE = it }
            }
        }
    }
    
    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()
    
    private val prefs: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    
    /**
     * Fetch the next upcoming shift for the authenticated user
     */
    suspend fun fetchNextShift(): ShiftData? = withContext(Dispatchers.IO) {
        val accessToken = prefs.getString(KEY_ACCESS_TOKEN, null)
        val userId = prefs.getString(KEY_USER_ID, null)
        
        if (accessToken.isNullOrEmpty() || userId.isNullOrEmpty()) {
            throw IllegalStateException("User not authenticated")
        }
        
        val now = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }.format(Date())
        
        val url = "$SUPABASE_URL/rest/v1/shifts" +
                "?owner_id=eq.$userId" +
                "&start_time=gt.$now" +
                "&order=start_time.asc" +
                "&limit=1"
        
        val request = Request.Builder()
            .url(url)
            .addHeader("apikey", SUPABASE_ANON_KEY)
            .addHeader("Authorization", "Bearer $accessToken")
            .addHeader("Content-Type", "application/json")
            .get()
            .build()
        
        try {
            val response = httpClient.newCall(request).execute()
            
            if (!response.isSuccessful) {
                throw IOException("HTTP ${response.code}: ${response.message}")
            }
            
            val responseBody = response.body?.string() ?: throw IOException("Empty response body")
            val jsonArray = JSONArray(responseBody)
            
            if (jsonArray.length() == 0) {
                return@withContext null
            }
            
            val shiftJson = jsonArray.getJSONObject(0)
            return@withContext parseShiftFromJson(shiftJson)
            
        } catch (e: Exception) {
            throw IOException("Failed to fetch shifts: ${e.message}", e)
        }
    }
    
    /**
     * Store authentication data for widget use
     */
    fun storeAuthenticationData(accessToken: String, userId: String) {
        prefs.edit()
            .putString(KEY_ACCESS_TOKEN, accessToken)
            .putString(KEY_USER_ID, userId)
            .apply()
    }
    
    /**
     * Clear stored authentication data
     */
    fun clearAuthenticationData() {
        prefs.edit()
            .remove(KEY_ACCESS_TOKEN)
            .remove(KEY_USER_ID)
            .apply()
    }
    
    /**
     * Check if user is authenticated
     */
    fun isAuthenticated(): Boolean {
        val accessToken = prefs.getString(KEY_ACCESS_TOKEN, null)
        val userId = prefs.getString(KEY_USER_ID, null)
        return !accessToken.isNullOrEmpty() && !userId.isNullOrEmpty()
    }
    
    private fun parseShiftFromJson(json: JSONObject): ShiftData {
        val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }
        
        return ShiftData(
            id = json.getString("id"),
            title = json.getString("title"),
            startTime = dateFormat.parse(json.getString("start_time").replace("Z", "").substring(0, 19)) ?: Date(),
            endTime = dateFormat.parse(json.getString("end_time").replace("Z", "").substring(0, 19)) ?: Date(),
            location = if (json.isNull("location")) null else json.getString("location"),
            description = if (json.isNull("description")) null else json.getString("description")
        )
    }
}

File: ios/SkiftappenWidget/SupabaseService.swift
import Foundation

// MARK: - Supabase Service for Widget
class SupabaseService {
    static let shared = SupabaseService()
    
    private let supabaseURL = "https://fsefeherdbtsddqimjco.supabase.co"
    private let supabaseAnonKey = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImZzZWZlaGVyZGJ0c2RkcWltamNvIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NTI3ODUwNDcsImV4cCI6MjA2ODM2MTA0N30.YEltOJVQU6Ox5YrkZJGzbMiojyQClkFwG-mBPilIAfk"
    
    private init() {}
    
    // MARK: - Fetch User Shifts
    func fetchUserShifts() async throws -> [ShiftData] {
        // Get user token from UserDefaults (should be set by main app)
        guard let userToken = UserDefaults(suiteName: "group.skiftappen")?.string(forKey: "supabase_access_token"),
              let userId = UserDefaults(suiteName: "group.skiftappen")?.string(forKey: "user_id") else {
            throw SupabaseError.notAuthenticated
        }
        
        // Prepare URL
        guard let url = URL(string: "\(supabaseURL)/rest/v1/shifts") else {
            throw SupabaseError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("Bearer \(supabaseAnonKey)", forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(userToken)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        // Add query parameters
        var components = URLComponents(url: url, resolvingAgainstBaseURL: false)!
        components.queryItems = [
            URLQueryItem(name: "owner_id", value: "eq.\(userId)"),
            URLQueryItem(name: "start_time", value: "gt.\(ISO8601DateFormatter().string(from: Date()))"),
            URLQueryItem(name: "order", value: "start_time.asc"),
            URLQueryItem(name: "limit", value: "5")
        ]
        
        request.url = components.url
        
        // Perform request
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw SupabaseError.invalidResponse
        }
        
        guard httpResponse.statusCode == 200 else {
            throw SupabaseError.httpError(httpResponse.statusCode)
        }
        
        // Parse response
        let shifts = try JSONDecoder().decode([SupabaseShift].self, from: data)
        
        return shifts.map { shift in
            ShiftData(
                id: shift.id,
                title: shift.title,
                startTime: ISO8601DateFormatter().date(from: shift.start_time) ?? Date(),
                endTime: ISO8601DateFormatter().date(from: shift.end_time) ?? Date(),
                location: shift.location,
                description: shift.description,
                isPlaceholder: false,
                isError: false,
                isEmpty: false
            )
        }
    }
    
    // MARK: - Store Authentication Data
    func storeAuthenticationData(accessToken: String, userId: String) {
        let userDefaults = UserDefaults(suiteName: "group.skiftappen")
        userDefaults?.set(accessToken, forKey: "supabase_access_token")
        userDefaults?.set(userId, forKey: "user_id")
    }
    
    // MARK: - Clear Authentication Data
    func clearAuthenticationData() {
        let userDefaults = UserDefaults(suiteName: "group.skiftappen")
        userDefaults?.removeObject(forKey: "supabase_access_token")
        userDefaults?.removeObject(forKey: "user_id")
    }
}

// MARK: - Data Models
struct SupabaseShift: Codable {
    let id: String
    let owner_id: String
    let start_time: String
    let end_time: String
    let title: String
    let description: String?
    let location: String?
    let created_at: String
    let updated_at: String
}

// MARK: - Errors
enum SupabaseError: Error, LocalizedError {
    case notAuthenticated
    case invalidURL
    case invalidResponse
    case httpError(Int)
    
    var errorDescription: String? {
        switch self {
        case .notAuthenticated:
            return "User not authenticated"
        case .invalidURL:
            return "Invalid URL"
        case .invalidResponse:
            return "Invalid response"
        case .httpError(let code):
            return "HTTP error: \(code)"
        }
    }
}